### DISCUSSION TITLE
SSHOperator exit code

### DISCUSSION THREAD

**Discussion by kolfild26** [apache/airflow#23788] [Q&A] [answered]
Hi,

I'm using `SSHOperator` to run bash scripts in the remote server. I wonder what is the best way to retrive the bash script (or just set of commands) exit code.
Assume, we have the script named `run.sh`:

```
#!/bin/bash

#do something
echo "something's done"

exit 123
```
Next we write the task which uses `SSHOperator`:
```
SSHOperator(
        task_id=ssh_task,
        command='path_to/run.sh',
        ssh_conn_id='my_ssh_conn)
```
As we know all the stuff which we `echo` on run.sh will go to XComs's return_value. In case of success, it'll be - "something's done". In case of fail - nothing at all.
**There is no exit code.**

- Add `echo $?` to the end of command and we will have return_value="something's done123" in XCom:
`command='path_to/run.sh || echo $?'`
- Suppress `>/dev/null` all other stuff except `$?` we get exactly what we need:
`command='path_to/run.sh >/dev/null || echo $?'`
return_value=123

It works, but missing some things.
The way I described above makes the task always finish successfully, as `echo $?` always  finishes successfully, It forces to add one more dummy/python task **for each ssh task** to analyze the return_value of ssh.

I found the only discussion regarding the same and there was [a good point](https://github.com/apache/airflow/discussions/19312#discussioncomment-1569048) by @potiuk  

> > Marking task as FAILED
> 
> What else did you expect to happen ?

So, my answer - In fail scenario I'd like the exit code to go to XCom without that workaround of making the ssh task always Ok to push the stuff to Xcom.
**Now, in case of  fail scenario nothing goes to XCom and therefore nothing to analyze in dag logic.**

**Question:**
Is there any way to achive two things simultaneously:

- task fails
- return_value=ssh command exit code

**Comment by potiuk** [comment]
No. But you can add it as a PR. That woudl be a nice improvement I think.

You can - even now in your DAG - simply extend SSHOperator and change it's behaviour in your DAGs:- this is entirely OK in Airlfow  - you should not treat the operators as "black-boxes" - it's perfectly fine to write your own operator extending those from Airflow that miss some functionality you would like to do. Same as writing DAG, writing an operator or extending existing one is usually few lines of Python code. It's as easy as this:

```
class MySSHOperator(SSHOperator):
     .... # Here you add modified methods that do what you want
```

You  can override the methods that get the exit_status and push them to Xcom. 

Then - once you test it locally, you are even most welcome to contribute it back as long as you make it reusable, configurable and backwards-compatible.

**Comment by kolfild26** [comment]
thanks, I know about rewriting the operators and even tried before. But as always, quick search for any already existed solution is the first step)

Let me ask more.

It was easy to find a place where `exit_code` is being analyzed. It is [there](https://github.com/apache/airflow/blob/main/airflow/providers/ssh/operators/ssh.py#L151).
```
    def raise_for_status(self, exit_status: int, stderr: bytes) -> None:
        if exit_status != 0:
            error_msg = stderr.decode('utf-8')
            raise AirflowException(f"error running cmd: {self.command}, error: {error_msg}")
```
It's binary (zero or not) as expected. Then it's possible to handle this. But next is still unclear. Niether [/hooks/ssh.py](https://github.com/apache/airflow/blob/main/airflow/providers/ssh/hooks/ssh.py) nor [/operators/ssh.py](https://github.com/apache/airflow/blob/main/airflow/providers/ssh/operators/ssh.py) use XCom directly.
It works so that if the execute method returns something, then it's passed to XCom. Else (AirflowException), nothing's passed to XCom
```
        try:
            with self.get_ssh_client() as ssh_client:
                result = self.run_ssh_client_command(ssh_client, self.command)
        except Exception as e:
            raise AirflowException(f"SSH operator error: {str(e)}")
        return result
```
I mean it goes beyond the SSHOperator's logic only. Here, the operator just returns or not.

> You can override the methods that get the exit_status and push them to Xcom.

So, you mean directly using `airflow.models.xcom.XCom` [set()](https://github.com/apache/airflow/blob/main/airflow/models/xcom.py#L154) method somewhere to push exit_status? Well, I'll try. If I got you correctly.

**Reply by kolfild26** [reply]
> It works so that if the execute method returns something, then it's passed to XCom. Else (AirflowException), nothing's passed to XCom

Ok, It seems the [task instance](https://github.com/apache/airflow/blob/54aa23405240b9db553bf639de85b2d86364420b/airflow/models/taskinstance.py#L2281) does all further things.

**Reply by potiuk** [reply]
Yeah. you just need to do xcom_push() wit the value. There are many operators that use it to push extra values.

**Comment by kolfild26** [comment, marked as answer]
Just posting here the solution.

Exit status is available in [def raise_for_status](https://github.com/apache/airflow/blob/22a0b517dd139fe8ab5bf699ca2990142787f199/airflow/providers/ssh/operators/ssh.py#L150). So, **fisrt**, need to pass the context to this method to further get the task instance.

[result = self.run_ssh_client_command(ssh_client, self.command)](https://github.com/apache/airflow/blob/22a0b517dd139fe8ab5bf699ca2990142787f199/airflow/providers/ssh/operators/ssh.py#L173) changed to:
`result = self.run_ssh_client_command(ssh_client, self.command, context=context)`

[self.raise_for_status(exit_status, agg_stderr)](https://github.com/apache/airflow/blob/22a0b517dd139fe8ab5bf699ca2990142787f199/airflow/providers/ssh/operators/ssh.py#L160) to:
`self.raise_for_status(exit_status, agg_stderr, context=kwargs['context'])`

[def raise_for_status(self, exit_status: int, stderr: bytes) -> None:](https://github.com/apache/airflow/blob/22a0b517dd139fe8ab5bf699ca2990142787f199/airflow/providers/ssh/operators/ssh.py#L150) to:
`def raise_for_status(self, exit_status: int, stderr: bytes, **kwargs) -> None:`

**Second**, add the next lines to the raise_for_status method
```
        ti=kwargs['context'].get('task_instance')
        ti.xcom_push(key='SSH Exit', value=exit_status)
```
Full code of raise_for_status:
```
    def raise_for_status(self, exit_status: int, stderr: bytes, **kwargs) -> None:
        ti=kwargs['context'].get('task_instance')
        ti.xcom_push(key='SSH Exit', value=exit_status)
        if exit_status != 0:
            error_msg = stderr.decode('utf-8')
            raise AirflowException(f"error running cmd: {self.command}, error: {error_msg}")
```
In both scenarios (success/failure) it will write the exit code to XCom **SSH Exit** variable

**Reply by potiuk** [reply]
Again - might be a good idea to contrbute it back :)

**Reply by kolfild26** [reply]
https://github.com/apache/airflow/pull/27370
